src/core/meshcop/border_agent.hpp

```diff
@@ -146,9 +151,20 @@ class BorderAgent : public InstanceLocator, private NonCopyable
     uint16_t GetUdpPort(void) const;
 
     /**
-     * Starts the Border Agent service.
+     * Starts the Border Agent service on the default port.
      */
-    void Start(void) { IgnoreError(Start(kUdpPort)); }
+    void Start(void) { IgnoreError(Start(kDefaultUdpPort)); }
+
+    /**
+     * Starts the Border Agent service on the specified port.
+     *
+     * @param[in] aUdpPort     The UDP port to listen on. If UDP port is zero, a dynamic port will be
+     *                         used. `GetUdpPort()` will return the current UDP port being used.
+     *
+     * @retval kErrorNone  Successfully started the Border Agent.
+     * @retval ...         Failed to start the Border Agent (e.g., could not bind to UDP port).
+     */
+    Error Start(uint16_t aUdpPort);
```

Comment:
Thanks, @milas.

For the use case of allowing the port number to be specified, we don't need to make the `Start()` method `public`. The Border Agent itself will start/stop based on events (e.g., when attached/detached, etc.).

Note that #11087 is making `Start()` and `Stop()` private. They were exposed as `public` to allow a workaround solution due to sharing the underlying DTLS `SecureAgent` between multiple modules. The `Commissioner` module would stop and restart the `BorderAgent` when it was itself enabled/disabled. This was more of a hack to allow certain tests to be performed when the two features were used together. The PR above is removing this hack.

Regarding your use case, here is my suggestion:

- We can add a new method, `Error SetUdpPort(uint16_t aUdpPort)`, which sets the default UDP port to be used by BA. We should restrict this to when BA is stopped (return an error otherwise).
- We track the default in a new member variable, `mDefaultUdpPort`, which we initialize with `kUdpPort` from the `OPENTHREAD_CONFIG_BORDER_AGENT_UDP_PORT`.
- This new API can then be exposed in `openthread/border_agent.h` as well (as `otBorderAgentSetUdpPort(...)`).
- I would also suggest adding a corresponding CLI command for this (e.g., update `@cli ba port` to allow both get and set).